x/shield/keeper/proposal.go

```diff
@@ -73,48 +97,65 @@ func (k Keeper) LockProvider(ctx sdk.Context, delAddr sdk.AccAddress, amount sdk
 	provider.Collateral = provider.Collateral.Sub(amount)
 	k.SetProvider(ctx, delAddr, provider)
 
-	// if there are enough delegations
-	// TODO logics for providers are changing, check if it outdated
+	// if there are enough delegations, do nothing
 	if provider.DelegationBonded.GTE(provider.TotalLocked) {
 		return
 	}
 
 	// if there are not enough delegations, check unbondings
 	unbondingDelegations := k.GetSortedUnbondingDelegations(ctx, delAddr)
-	short := provider.TotalLocked.Sub(provider.DelegationBonded)
+	remaining := provider.TotalLocked.Sub(provider.DelegationBonded)
 	endTime := ctx.BlockTime().Add(lockPeriod)
 	for _, ubd := range unbondingDelegations {
-		if !short.IsPositive() {
+		if !remaining.IsPositive() {
 			return
 		}
-		for i, entry := range ubd.Entries {
-			if entry.CompletionTime.Before(endTime) {
-				// change unbonding completion time
-				timeSlice := k.sk.GetUBDQueueTimeSlice(ctx, ubd.Entries[i].CompletionTime)
+		entry := ubd.Entries[0]
+		if entry.CompletionTime.Before(endTime) {
+			// change unbonding completion time
+			timeSlice := k.sk.GetUBDQueueTimeSlice(ctx, entry.CompletionTime)
+			if len(timeSlice) > 1 {
 				for i := 0; i < len(timeSlice); i++ {
 					if timeSlice[i].DelegatorAddress.Equals(delAddr) && timeSlice[i].ValidatorAddress.Equals(ubd.ValidatorAddress) {
 						timeSlice = append(timeSlice[:i], timeSlice[i+1:]...)
 						k.sk.SetUBDQueueTimeSlice(ctx, entry.CompletionTime, timeSlice)
 						break
 					}
 				}
-				ubd.Entries[i].CompletionTime = endTime
-				k.sk.InsertUBDQueue(ctx, ubd, endTime)
+			} else {
+				ctx.KVStore(k.stakingStoreKey).Delete(staking.GetUnbondingDelegationTimeKey(entry.CompletionTime))
```

Comment:
Can we just delete it without having direct staking store key access? I think this is a really bad practice in general. I'd suggest we wrap staking keeper and implement a function instead of having this

Reply:
Yeah actually I didn't want to do this also, but the staking module does not provide a method to delete it. Actually leaving an empty element there will not cause any problem in real run. It's a problem with staking import/export simulation. The simulation will check if the stores before and after import/export are the same. It does not require queues (unbonding, redelegation and validator) to be the same, but the numbers of key-value pairs should be the same. If there's an empty element in the queue, it does not cause any problem but will fail the import/export simulation.

Reply:
I also tried to wrap staking keeper. I didn't finish it because the cosmos distribution module `stakingkeeper.Keeper` rather than an expected keeper is required. I'll try to figure out if if this problem can be solved.

Reply:
Fixed.